src/topology_utils.py

```diff
@@ -0,0 +1,278 @@
+
+def get_auth_session(args):
+    """
+    Return a requests session ready for an XML query.
+    """
+    euid = os.geteuid()
+    if euid == 0:
+        cert = '/etc/grid-security/hostcert.pem'
+        key = '/etc/grid-security/hostkey.pem'
+    else:
+        cert = '/tmp/x509up_u%d' % euid
+        key = '/tmp/x509up_u%d' % euid
+
+    cert = os.environ.get('X509_USER_PROXY', cert)
+    key = os.environ.get('X509_USER_PROXY', key)
+
+    if args.cert:
+        cert = args.cert
+    if args.key:
+        key = args.key
+
+    session = requests.Session()
+
+    if os.path.exists(cert):
+        session.cert = cert
+    if os.path.exists(key):
+        session.cert = (cert, key)
+
+    return session
+
+
+def update_url_hostname(url, args):
+    """
+    Given a URL and an argument object, update the URL's hostname
+    according to args.host and return the newly-formed URL.
+    """
+    if not args.host:
+        return url
+    url_tuple = list(urlparse.urlsplit(url))
+    url_tuple[1] = args.host
+    return urlparse.urlunsplit(url_tuple)
+
+
+def get_contact_list_info(contact_list):
+    """
+    Get contact list info out of contact list
+    """
+    contact_list_info = []
+    for contact in contact_list:
+        if contact.tag == 'ContactType' or contact.tag == 'Type':
+            contact_list_type = contact.text.lower()
+        if contact.tag == 'Contacts':
+            for contact in contact:
+                contact_info = { 'ContactType' : contact_list_type }
+                for contact_contents in contact:
+                    contact_info[contact_contents.tag] = contact_contents.text
+                contact_list_info.append(contact_info)
+
+    return contact_list_info
+
+
+def get_vo_map(args, session=None):
+    """
+    Generate a dictionary mapping from the VO name (key) to the
+    VO ID (value).
+    """
+    url = update_url_hostname("https://my.opensciencegrid.org/vosummary"
+                              "/xml?all_vos=on&active_value=1", args)
+    if session is None:
+        with get_auth_session(args) as session:
+            response = session.get(url)
+    else:
+        response = session.get(url)
+
+    if response.status_code != requests.codes.ok:
+        raise Exception("MyOSG request failed (status %d): %s" % \
+              (response.status_code, response.text[:2048]))
+
+    root = ET.fromstring(response.content)
+    if root.tag != 'VOSummary':
+        raise Exception("MyOSG returned invalid XML with root tag %s" % root.tag)
+    vo_map = {}
+    for child_vo in root:
+        if child_vo.tag != "VO":
+            raise Exception("MyOSG returned a non-VO  (%s) inside VO summary." % \
+                            root.tag)
+        vo_info = {}
+        for child_info in child_vo:
+            vo_info[child_info.tag] = child_info.text
+        if 'ID' in vo_info and 'Name' in vo_info:
+            vo_map[vo_info['Name'].lower()] = vo_info['ID']
+
+    return vo_map
+
+
+SERVICE_IDS = {'ce': 1,
```

Comment:
You can get service IDs from `topology/services.yaml`

Reply:
Hm.  I'd prefer to keep this PR to refactoring.  I'll open a new issue for this idea.